mockwebserver/src/main/java/okhttp3/mockwebserver/MockWebServer.kt

```diff
@@ -0,0 +1,1114 @@
+      sink.flush()
+    }
+
+    var hasBody = false
+    val requestBody = TruncatingBuffer(bodyLimit)
+    val chunkSizes = ArrayList<Int>()
+    val policy = dispatcher.peek()
+    if (contentLength != -1L) {
+      hasBody = contentLength > 0
+      throttledTransfer(policy, socket, source, requestBody.buffer(), contentLength, true)
+    } else if (chunked) {
+      hasBody = true
+      while (true) {
+        val chunkSize = Integer.parseInt(source.readUtf8LineStrict().trim { it <= ' ' }, 16)
+        if (chunkSize == 0) {
+          readEmptyLine(source)
+          break
+        }
+        chunkSizes.add(chunkSize)
+        throttledTransfer(policy, socket, source,
+            requestBody.buffer(), chunkSize.toLong(), true)
+        readEmptyLine(source)
+      }
+    }
+
+    val method = request.substring(0, request.indexOf(' '))
+    if (hasBody && !HttpMethod.permitsRequestBody(method)) {
+      throw IllegalArgumentException("Request must not have a body: $request")
+    }
+
+    return RecordedRequest(request, headers.build(), chunkSizes, requestBody.receivedByteCount,
+        requestBody.buffer, sequenceNumber, socket)
+  }
+
+  @Throws(IOException::class)
+  private fun handleWebSocketUpgrade(
+    socket: Socket,
+    source: BufferedSource,
+    sink: BufferedSink,
+    request: RecordedRequest,
+    response: MockResponse
+  ) {
+    val key = request.getHeader("Sec-WebSocket-Key")
+    response.setHeader("Sec-WebSocket-Accept", WebSocketProtocol.acceptHeader(key!!))
+
+    writeHttpResponse(socket, sink, response)
+
+    // Adapt the request and response into our Request and Response domain model.
+    val scheme = if (request.tlsVersion != null) "https" else "http"
+    val authority = request.getHeader("Host") // Has host and port.
+    val fancyRequest = Request.Builder()
+        .url("$scheme://$authority/")
+        .headers(request.headers)
+        .build()
+    val statusParts = response.getStatus().split(" ".toRegex(), 3).toTypedArray()
+    val fancyResponse = Response.Builder()
+        .code(Integer.parseInt(statusParts[1]))
+        .message(statusParts[2])
+        .headers(response.getHeaders())
+        .request(fancyRequest)
+        .protocol(Protocol.HTTP_1_1)
+        .build()
+
+    val connectionClose = CountDownLatch(1)
+    val streams = object : RealWebSocket.Streams(false, source, sink) {
+      override fun close() = connectionClose.countDown()
+    }
+    val webSocket = RealWebSocket(fancyRequest,
+        response.webSocketListener, SecureRandom(), 0)
+    response.webSocketListener!!.onOpen(webSocket, fancyResponse)
+    val name = "MockWebServer WebSocket ${request.path!!}"
+    webSocket.initReaderAndWriter(name, streams)
+    try {
+      webSocket.loopReader()
+
+      // Even if messages are no longer being read we need to wait for the connection close signal.
+      try {
+        connectionClose.await()
+      } catch (e: InterruptedException) {
+        throw AssertionError(e)
```

Comment:
whenever we have a catch clause that throws an `AssertionError`, we should just omit the `try/catch` entirely. Throwing an AssertionError is our way of saying “we don’t intend to handle this”

Reply:
Fixed.